From 02778ef75050509ca42f0fc59ab6598b28418d4b Mon Sep 17 00:00:00 2001
From: Julien Thomas <jthomas@zenetys.com>
Date: Sun, 16 Aug 2020 17:09:56 +0200
Subject: [PATCH] build: create static library and make it build on el8
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Inspired by Gentoo patch "dev-libs/grok: Rev bump to allow building
against >=gperf-3.1" from Thomas Deutschmann <whissi@gentoo.org>,
referenced on Github jordansissel/grok issue #29 "Failes to compile
with gperf 3.1 (conflicting types for ‘patname2macro’)".
---
 Makefile               | 29 +++++++++++++++++------------
 grok_matchconf_macro.h |  2 +-
 2 files changed, 18 insertions(+), 13 deletions(-)

diff --git a/Makefile b/Makefile
index aa60dba..57f4483 100644
--- a/Makefile
+++ b/Makefile
@@ -25,10 +25,14 @@ else
 GPERF?=/usr/bin/gperf
 endif
 
+LIBS=-lpcre -levent -rdynamic -ltokyocabinet
+
 # For linux, we need libdl for dlopen()
 # On FreeBSD, comment this line out.
 ifeq ($(PLATFORM), GNULinux)
-LDFLAGS+=-ldl
+LIBS+=-ldl
+# tirpc el8
+LIBS+=-ltirpc
 endif
 
 # #############################################
@@ -41,8 +45,7 @@ VERSION=$(shell sh $(BASE)/version.sh)
 #CFLAGS+=-g
 #LDFLAGS+=-g
 
-CFLAGS+=-pipe -fPIC -I. -O2
-LDFLAGS+=-lpcre -levent -rdynamic -ltokyocabinet
+CFLAGS+=-fPIC -I.
 
 LIBSUFFIX=$(shell sh $(BASE)/platform.sh libsuffix)
 VERLIBSUFFIX=$(shell sh $(BASE)/platform.sh libsuffix $(MAJOR))
@@ -53,6 +56,9 @@ LIBNAMEFLAG=$(shell sh $(BASE)/platform.sh libnameflag $(MAJOR) $(INSTALLLIB))
 CFLAGS+=-I/usr/local/include
 LDFLAGS+=-L/usr/local/lib
 
+# rpc/rpc.h on el8
+CFLAGS+=-I/usr/include/tirpc
+
 # Platform so we know what to dlopen
 CFLAGS+=-DPLATFORM_$(PLATFORM)
 # Uncomment to totally disable logging features
@@ -66,7 +72,7 @@ LDFLAGS+=$(EXTRA_LDFLAGS)
 ### End of user-servicable configuration
 
 CLEANGEN=filters.c grok_matchconf_macro.c *.yy.c *.tab.c *.tab.h
-CLEANOBJ=*.o *_xdr.[ch] *.so
+CLEANOBJ=*.o *_xdr.[ch] *.a *.so *.so.[0-9]
 CLEANBIN=main grokre grok conftest grok_program
 CLEANVER=VERSION grok_version.h grok.spec
 
@@ -83,7 +89,7 @@ GROKHEADER=grok.h grok_pattern.h grok_capture.h grok_capture_xdr.h \
 GROKHEADER+=grok_version.h
 
 .PHONY: all
-all: grok discogrok libgrok.$(LIBSUFFIX) libgrok.$(VERLIBSUFFIX)
+all: grok discogrok libgrok.a libgrok.$(LIBSUFFIX) libgrok.$(VERLIBSUFFIX)
 
 .PHONY: package create-package test-package update-version
 package: 
@@ -162,14 +168,17 @@ cleanver:
 # Binary creation
 grok: LDFLAGS+=-levent
 grok: $(GROKOBJ) conf.tab.o conf.yy.o main.o grok_config.o
-	$(CC) $^ -o $@ $(LDFLAGS) 
+	$(CC) $(LDFLAGS) $^ $(LIBS) -o $@
 
 discogrok: $(GROKOBJ) discover_main.o
-	$(CC) $^ -o $@ $(LDFLAGS)
+	$(CC) $(LDFLAGS) $^ $(LIBS) -o $@
+
+libgrok.a: $(GROKOBJ)
+	ar rcs $@ $<
 
 libgrok.$(LIBSUFFIX): 
 libgrok.$(LIBSUFFIX): $(GROKOBJ) 
-	$(CC) -fPIC $(DYNLIBFLAG) $(LIBNAMEFLAG) $^ -o $@ $(LDFLAGS) 
+	$(CC) $(LDFLAGS) -fPIC $(DYNLIBFLAG) $(LIBNAMEFLAG) $^ $(LIBS) -o $@
 
 libgrok.$(VERLIBSUFFIX): libgrok.$(LIBSUFFIX);
 	ln -s $< $@
@@ -213,10 +222,6 @@ grok_capture_xdr.h: grok_capture.x
 	rpcgen -h $< -o $@
 
 %.c: %.gperf
-	@if $(GPERF) --version | head -1 | egrep -v '3\.[0-9]+\.[0-9]+' ; then \
-		echo "We require gperf version >= 3.0.3" ; \
-		exit 1; \
-	fi
 	$(GPERF) $< > $@
 
 conf.tab.c conf.tab.h: conf.y
diff --git a/grok_matchconf_macro.h b/grok_matchconf_macro.h
index 85876f6..f1fe1f3 100644
--- a/grok_matchconf_macro.h
+++ b/grok_matchconf_macro.h
@@ -19,6 +19,6 @@ struct strmacro {
 #endif
 
 /* this function is generated by gperf */
-const struct strmacro *patname2macro(const char *str, unsigned int len);
+const struct strmacro *patname2macro(const char *str, size_t len);
 
 #endif /* _GROK_MATCHCONF_MACRO_ */
-- 
2.21.1

